SDK bump, new ad units, separate load and show for interstitials

RedBrogdon opened the pull request:
Hello from the AdMob DevRel team!

I'm making some updates to the firebase-admob plugin as a first step toward teaching myself Flutter. This is my first time coding in Dart and my first Flutter-related commit, so feel free to critique anything. :)

RedBrogdon commented:
TIL about proper line length for .dart files.

RedBrogdon commented:
The remaining Travis failure is for the iOS build:

> Found "Xcode 8.3.3". Xcode 9.0 or greater is required to develop for iOS.
Encountered error while building for device.

I don't see this error for any of the other PRs, but I also haven't touched anything specific to iOS here. Any advice?

lukef commented:
@RedBrogdon - you may want to bump gradle to 3.0.1 as that is the latest. i.e.: `com.android.tools.build:gradle:3.0.1`

RedBrogdon commented:
Nice catch. Done.

lukef commented:
re: xcode. It might be related to cocoapods. The example does not specify a `platform` value but travis is set up to use xcode 8.3.

The example app should probably specify a `platform` value and the travis configuration should also probably be updated to use xcode 9.

"CocoaPods provides a default deployment target if one is not specified. The current default values are 4.3 for iOS, 10.6 for OS X, 9.0 for tvOS and 2.0 for watchOS."

RedBrogdon commented:
Hrm. Xcode is still failing, and I don't want to try messing with the Travis configuration for a project for which I'm not yet a contributor

If anyone from the Flutter team can point me toward the best way to resolve this, much obliged.  I'll hold here and work on some other changes locally for now.

zanderso commented:
/cc @mravn-google

mravn-google commented:
We're currently seeing the same Xcode 8 vs Xcode 9 issue for all plugins (I believe) due to https://github.com/flutter/flutter/commit/7fb78523e9efe422de60a9076831f753b7f69e31. We seem to need an upgrade to the Travis config for flutter/plugins. I'm on it.

mravn-google commented on packages/firebase_admob/android/build.gradle:
```diff
@@ -30,7 +36,8 @@ android {
-        compile 'com.google.firebase:firebase-core:11.4.+'
-        compile 'com.google.firebase:firebase-ads:11.4.+'
+        // firebase-core is exposed so the google-services plugin will see it.
+        api 'com.google.firebase:firebase-core:11.6.2'
```
I'm a bit uneasy about pinning these dependencies to specific versions. Many Flutter apps are going to be using several Firebase-related plugins together, and then they face the challenge of ensuring consistency among transient dependencies like `firebase-core`. Since the other plugins currently (and unfortunately) use `11.4.+`, we have an inconsistency with `11.6.2`. Is upgrading to `11.6.2` necessary? If not, please stay at `11.4.+`. Otherwise, we need to provide consistent versions of the other Firebase plugins as well. Those updates could then bump the Gradle versions as well. And I'd suggest using the version range specifier `[11.6.2,12.0[` to avoid a similar issue in the future, at least until somebody needs v12.

Skylled replied:
Cloud Firestore also needs a bump to `11.6.+`, [see comment](https://github.com/flutter/plugins/pull/282#issuecomment-350100213).

RedBrogdon replied:
<facepalm> Right, this plugin doesn't exist in a vacuum, and is instead part of a suite that needs to work together. I should have accounted for that.

I don't need any recently released features, so there's no need to update the versions in this file now. When they *do* get updated, it should clearly be in concert with the others. Change reverted.

RedBrogdon commented on packages/firebase_admob/android/build.gradle:
```diff
@@ -3,25 +3,31 @@ version '1.0-SNAPSHOT'
 
 buildscript {
     repositories {
+        maven {
```
Done.

RedBrogdon commented on packages/firebase_admob/android/build.gradle:
```diff
@@ -3,25 +3,31 @@ version '1.0-SNAPSHOT'
 
     dependencies {
-        classpath 'com.android.tools.build:gradle:2.3.0'
+        classpath 'com.android.tools.build:gradle:3.0.1'
```
Reverted. See my comment below on the firebase-core and -ads versions.

RedBrogdon commented on packages/firebase_admob/android/build.gradle:
```diff
@@ -3,25 +3,31 @@ version '1.0-SNAPSHOT'
 
 rootProject.allprojects {
     repositories {
+        maven {
```
Done.

RedBrogdon commented on packages/firebase_admob/android/build.gradle:
```diff
@@ -3,25 +3,31 @@ version '1.0-SNAPSHOT'
-    compileSdkVersion 25
-    buildToolsVersion '25.0.3'
+    compileSdkVersion 26
+    buildToolsVersion '26'
```
Done.

RedBrogdon commented on packages/firebase_admob/example/android/app/build.gradle:
```diff
@@ -15,8 +15,8 @@ apply plugin: 'com.android.application'
-    compileSdkVersion 25
-    buildToolsVersion '25.0.3'
+    compileSdkVersion 26
+    buildToolsVersion '26'
```
Done.

RedBrogdon commented on packages/firebase_admob/example/android/app/build.gradle:
```diff
@@ -40,10 +40,16 @@ flutter {
-    androidTestCompile 'com.android.support:support-annotations:25.2.0'
-    androidTestCompile 'com.android.support.test:runner:0.5'
-    androidTestCompile 'com.android.support.test:rules:0.5'
+    androidTestImplementation 'com.android.support:support-annotations:26.1.0'
```
Done.

RedBrogdon commented on packages/firebase_admob/example/ios/Podfile:
```diff
@@ -1,5 +1,5 @@
 # Uncomment this line to define a global platform for your project
-# platform :ios, '9.0'
+platform :ios, '8.0'
```
This was just an attempt to try to unbreak the Travis Xcode test. It's reverted now.

RedBrogdon commented:
Thanks for the quick feedback! I believe I've addressed your concerns, and with the Travis config change 9.2 is building the iOS version just fine. Let me know if I missed anything or there are other issues, though.

On a related note, I've begun adding rewarded video ads to the plugin in another local branch. Would it be a normal practice to send a work-in-progress PR with just the Dart code (no iOS or Android platform code), so someone can sanity check it and the API before I keep going?

mravn-google commented:
@RedBrogdon Sounds great! If you create a PR that represents work in progress, just mark it as such in the title (e.g. [WIP] Xxx or [Review only] Yyy) and add a comment saying what you'd like reviewers to focus on and what to ignore for now.

lukef commented:
@RedBrogdon @mravn-google - It might also make sense to bump the compile + target sdk to `27` and build tools to `27.0.2` as they are the latest versions. Also, I know we're using inclusive dependency versioning, but it's worth noting that Google Play Services 11.8 came out today as well so all firebase deps will use 11.8.

mravn-google commented:
@lukef Our current strategy is to follow Android Studio templates as much as possible/reasonable. And those are still at v26, I believe (at least, that is the case on my machine). Not sure why, but possibly related to the fact that the SDK manager says v27 is only "partially installed", offering me no obvious way to complete it.

mravn-google commented:
@RedBrogdon @lukef FYI, https://github.com/flutter/plugins/pull/317 is on its way too, with Gradle and dependency upgrades to Firebase-related plugins.

lukef commented:
re: partially installed - you need to click the "Show Package Details" checkbox at the bottom of the list. This will show the full set of options.

![capture](https://user-images.githubusercontent.com/37941/34172383-b4b5d53c-e4a6-11e7-9d19-c0ca2e30a3fc.PNG)

mravn-google commented:
@RedBrogdon Sorry, we should get you in a position to land this. May I ask you to write a `CHANGELOG.md` entry and bump the version in `pubspec.yaml`? I'll deal with the version and template issues as part of https://github.com/flutter/plugins/pull/317.

RedBrogdon commented:
Sure thing. I'm doing some traveling for the holidays, but can have it for you in a day or two.

sethladd commented:
@RedBrogdon thank you so much for the contribution!

lukef commented:
Sorry to be a pest, but does this mean that flutter will be targeting SDK 26 as a maximum? this seems contrary to what we should be doing on the android native side.

mravn-google commented:
@lukef No. The project templates are not normative; they are intended to provide a starting point.

lukef commented:
@mravn-google - Happy to jump on a call to discuss, but .. the issue I see is that if you lock the plugin to `26` and (specifically) build-tools `26.0.2` and, lets say, you have only `26.0.1` installed then you will need to pull down v `26.0.2` on top also. so if plugin A is 26.0.1 and plugin B is 26.0.2 and you app is 27.0.2 then you have 3 versions in play.

I'm working on a loose proposal to work around it, but in the meantime my thoughts are if you push people to the highest version it is better. it'd be better for a developer to be forced to bump the build version in their app vs having to fight each plugin if they want to use 27.

lukef commented:
Full disclosure: I don't know fully the direct consequences of the above scenario other than it using a bunch of bandwidth and disk space. There are other side effects that are best shared in an email because its probably making this PR messy :)

RedBrogdon commented:
Oh, look at me forgetting to pull changes.

RedBrogdon commented:
Clearly I'm learning as much about developing on GitHub as I am about Dart and Flutter. It looks like I've doubled up commits while rebasing onto the changes from master, which has bloated the PR and made it look (to my untrained eye) impossible to review.

Is there a simple way to fix this that you can recommend, or should I just blow away this pull request and make a new, clean one?

RedBrogdon commented:
I'm going to go ahead and close this PR and create a new one. Should be ready in a jiffy.

RedBrogdon commented:
New pull requested submitted as #327.
